### PR TITLE
PHP 8.0 - Use strict comparison to compare post dates

### PR DESCRIPTION
**Pull request by james-allan**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [ ] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

PHP 8.0 contains breaking changes to non-strict comparisons between integers and strings where the string contains non-numeric characters. This affects comparisons like `0 < '0000-00-00 00:00:00'`

With PHP 8.0, that comparison equates to `true` whereas prior to 8.0 it would be `false`.

More details of this change can be found at https://wiki.php.net/rfc/string_to_number_comparison

Please let me know if you would rather a different comparison to be used here. While the new logic isn't equivalent, I couldn't imagine a circumstance where the post's GMT dates would be an integer or some other format where a looser comparison would be necessary but I'm in your hands on that. 

### Other information:

* [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [ ] Have you successfully run tests with your changes locally?

### Changelog entry

> Enter a summary of all changes on this Pull Request. This will appear in the changelog if accepted.

### CODE REVIEW

**Comment by Konamiman** [comment]
Hi @james-allan,

Thanks for your contribution. I'm working on ensuring PHP 8 compatibility for WooCommerce and I had missed that one. 🙂

Your approach seems safe, given that these dates always come from non-nullable, DateTime fields in the database; also we are already using direct comparison with `'0000-00-00 00:00:00'` in other parts of the code (see for example `WC_Webhook_Data_Store`).

However I'm a bit concerned about code duplication. Given that all three classes that you have patched ultimately inherit from `WC_Data_Store_WP`, do you think you could add an auxiliary method in that class? Something along the lines of this:

```php
protected function string_to_timestamp( string $time_string ) {
    return '0000-00-00 00:00:00' !== $time_string ? wc_string_to_timestamp( $time_string ) : null;
}
```

And while we're at it, there's another file where this wrong type of comparison is used, it would be great to fix it as well in this PR: [class-wc-coupon-data-store-cpt.php](https://github.com/woocommerce/woocommerce/blob/master/includes/data-stores/class-wc-coupon-data-store-cpt.php)

**Comment by james-allan** [comment]
Thanks for the feedback @Konamiman. I've introduced that function as you suggested and updated the coupon datastore class too. 🙂 

Let me know what you think.

File: includes/data-stores/class-wc-data-store-wp.php [hunk 1]
```diff
@@ -630,4 +630,16 @@ public function delete_from_lookup_table( $id, $table ) {
 		);
 		wp_cache_delete( 'lookup_table', 'object_' . $id );
 	}
+
+	/**
+	 * Converts a WP post date string into a timestamp.
+	 *
+	 * @since 4.7.0
```
Comment (Konamiman) [inline includes/data-stores/class-wc-data-store-wp.php:637]: This change won't make it to 4.7 so:

```suggestion
	 * @since 4.8.0
```

**Review by Konamiman — DISMISSED** [review]

**Review by Konamiman — APPROVED** [review]
Looks good. Approved!

**Pull request merged**